### PR TITLE
KAFKA-5409: Providing a custom client-id to the ConsoleProducer tool

### PR DESCRIPTION
**Pull request by ppatierno**

### CODE REVIEW

**Comment by ppatierno** [comment]
Retest this please

**Comment by ewencp** [comment]
@ppatierno Can you explain the goal of this? `client.id` is used to track resource usage on the broker, but if the ID is randomly generated it doesn't seem like there is a way to track resources back to an actual client?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6235/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6219/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ppatierno** [comment]
@ewencp we know that client.id is also useful for tracing calls and it identifies logically the application making a request to the broker. The main usage is about quotas as you said.
A lot of people who start working with Kafka try to interact with the broker using the console consumer and producer application (before writing a new application for their business). I think that they should have the possibility to specify the client.id for tracking/logging purposes and even for understanding how quotas work. In the current version, the "console-producer" is used and it's fixed so starting more than one console producer is impossible to distinguish them. Having a random client.id provides this and the user doesn't need to specify it explicitly. Maybe it could be useful to print the randomly generated client.id on application startup so that the user knows what to observe for tracing or for quotas ?

**Comment by ewencp** [comment]
I see, because that setting is applied after loading the user-specified configs, it's impossible to override currently. I think we could cut this patch down quite a bit just by handling that one case (i.e. only set to `console-producer` if the user didn't already specify a setting) and it accomplishes the same thing. I don't think it makes sense to assign random client IDs if the user hasn't specified one though.

**Comment by ppatierno** [comment]
@ewencp I have removed the random client.id generation if not specified (instead the fixed "console-producer" is used): I have also added another unit test when the client.id is not specified on the command line.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6303/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6287/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/tools/ConsoleProducer.scala [hunk 1]
```diff
@@ -96,23 +98,15 @@ object ConsoleProducer {
     props.put("serializer.class", config.valueEncoderClass)
     props.put("send.buffer.bytes", config.socketBuffer.toString)
     props.put("topic.metadata.refresh.interval.ms", config.metadataExpiryMs.toString)
-    props.put("client.id", "console-producer")
-
-    props
-  }
 
-  private def producerProps(config: ProducerConfig): Properties = {
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/ConsoleProducer.scala:104]: Hmm, so the refactoring here seems like it's going the opposite direction we'd want -- now there's unlabeled logic in main() that's really about the config and we've duplicated code for getting the user-specified configs included in both the old and new producer configs.

**Review by ewencp — COMMENTED** [review]
Just an organization nit, but otherwise this LGTM.

File: core/src/main/scala/kafka/tools/ConsoleProducer.scala (hunk 1 above)
Comment (ppatierno) [reply core/src/main/scala/kafka/tools/ConsoleProducer.scala:104]: It fixes the problem we had with overridden client.id property. If you see the ConsoleConsumer code now they are quite the same (the consumer case is related to handle group.id). Am I missing something here ? :(

**Comment by ppatierno** [comment]
@ewencp any news on that ?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6394/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6409/
Test PASSed (JDK 7 and Scala 2.11).

File: core/src/main/scala/kafka/tools/ConsoleProducer.scala (hunk 1 above)
Comment (ppatierno) [reply core/src/main/scala/kafka/tools/ConsoleProducer.scala:104]: @ewencp any new comments on that ?

Comment (ppatierno) [reply core/src/main/scala/kafka/tools/ConsoleProducer.scala:104]: @ewencp no news since August on this :(

**Comment by ppatierno** [comment]
@ewencp any more interest on this PR. Should I fix conflicts or just close this PR? Thanks!

**Pull request closed without merging**